### PR TITLE
[MNOE-1007] Improve Marketplace Filtering

### PR DESCRIPTION
**Pull request by NickenChuggets**
Created custom filter to only search certain fields
Moved filter to a ng-change method to improve performance

### CODE REVIEW

File: src/app/components/mno-apps/mno-app-listing.coffee [hunk 1]
```diff
@@ -23,12 +24,36 @@ angular.module 'mnoEnterpriseAngular'
...
-        if (vm.searchTerm? && vm.searchTerm.length > 0) || !vm.selectedCategory
-          return true
-        else
-          return _.contains(app.categories, vm.selectedCategory)
+      vm.onSearchChange = () ->
+        vm.selectedCategory = ''
+        vm.filteredApps = []
+        firstFilterResult = []
+        for app in vm.apps
+          if (vm.searchTerm? && vm.searchTerm.length > 0) || !vm.selectedCategory
```
Comment (yuritomanek) [inline src/app/components/mno-apps/mno-app-listing.coffee]: Use the existential operator here :

```
if vm.searchTerm?.length > 0 || !vm.selectedCategory
```

File: src/app/components/mno-apps/mno-app-listing.coffee [hunk 2]
```diff
@@ -23,12 +24,36 @@ angular.module 'mnoEnterpriseAngular'
...
+      vm.linkFor = (app) ->
+        "#/marketplace/#{app.id}"
 
-      vm.appsFilter = (app) ->
-        if (vm.searchTerm? && vm.searchTerm.length > 0) || !vm.selectedCategory
-          return true
-        else
-          return _.contains(app.categories, vm.selectedCategory)
+      vm.onSearchChange = () ->
+        vm.selectedCategory = ''
```
Comment (yuritomanek) [inline src/app/components/mno-apps/mno-app-listing.coffee]: I'm unsure why you're reseting `vm.selectedCategory` to `''`?
You're resetting it again on L52.
Just want to make sure this doesn't have any unintended effects.

File: src/app/components/mno-apps/mno-app-listing.coffee [hunk 3]
```diff
@@ -23,12 +24,36 @@ angular.module 'mnoEnterpriseAngular'
...
+        vm.filteredApps = []
+        firstFilterResult = []
+        for app in vm.apps
+          if (vm.searchTerm? && vm.searchTerm.length > 0) || !vm.selectedCategory
+            firstFilterResult.push(app)
+          else
+            if _.contains(app.categories, vm.selectedCategory)
+              firstFilterResult.push(app)
+        for app in firstFilterResult
+          if (vm.searchTerm? && vm.searchTerm.length > 0)
```
Comment (yuritomanek) [inline src/app/components/mno-apps/mno-app-listing.coffee]:
```
if (vm.searchTerm?.length > 0)
```

File: src/app/components/mno-apps/mno-app-listing.coffee [hunk 4]
```diff
@@ -27,12 +27,20 @@ angular.module 'mnoEnterpriseAngular'
       # Scope Management
       #====================================
 
+      vm.checktags = (tags, term) ->
+        res = false
+        index = 0
+        while !res and index < tags.length
+          res = tags[index]?.toLowerCase().indexOf(term) isnt -1
+          index++
+        res
```
Comment (ouranos) [inline src/app/components/mno-apps/mno-app-listing.coffee]: Or you could use lodash:
```coffeescript
_.some(tags, (item) -> item.toLowerCase().indexOf(term) isnt -1)
```

Comment (NickenChuggets) [reply src/app/components/mno-apps/mno-app-listing.coffee]: Done

**Pull request merged**